### DISCUSSION TITLE
`ReadableStream` in email message

### DISCUSSION THREAD

**Discussion by niquenen** [cloudflare/workerd#1501] [Q&A] [answered]
Currently with the Email Workers, an email message is defined with [`ForwardableEmailMessage`](https://github.com/cloudflare/workerd/blob/ef2c11f75d666c31faa7ab84b82eb4131a920d56/types/defines/email.d.ts#L18). The message contains a property `raw` with the type [`ReadableStream`](https://github.com/cloudflare/workerd/blob/ef2c11f75d666c31faa7ab84b82eb4131a920d56/types/defines/email.d.ts#L22) to read the message. [`ReadableStream`](https://github.com/cloudflare/workerd/blob/ef2c11f75d666c31faa7ab84b82eb4131a920d56/types/defines/email.d.ts#L22) is a generic type and the possible types are not defined.

<img width="577" alt="Screenshot 2023-12-21 at 11h56" src="https://github.com/cloudflare/workerd/assets/7528452/7c1b051c-d6fc-4be6-874b-c7cfed36e867">

With the [get started](https://developers.cloudflare.com/email-routing/email-workers/edit-email-workers/#add-an-email-worker) and the following script, every data read are an `Uint8Array` object.

```js
export default {
  /**
   * @param {{raw: ReadableStream}} message
   */
  async email(message, env, ctx) {
    const reader = message.raw.getReader();
    const read = await reader.read();
    if (typeof read == 'object') {
      console.log(`object type: ${read.value.constructor.name}`)
    }
    else {
      console.log(`type: ${typeof read}`)
    }
  }
}
```

Example of a log output from the Worker.

```json
{
  "outcome": "ok",
  "scriptName": "email-test",
  "diagnosticsChannelEvents": [],
  "exceptions": [],
  "logs": [
    {
      "message": [
        "object type: Uint8Array"
      ],
      "level": "log",
      "timestamp": 1703151398815
    }
  ],
  "eventTimestamp": 1703151398801,
  "event": {
    "rawSize": 11889,
    "rcptTo": "john.doe@example.com",
    "mailFrom": "from@example.com"
  },
  "id": 0
}
```

All types are accepted or the stream is only a `ReadableStream<Uint8Array>`?

**Comment by jasnell** [comment, marked as answer]
In this case, the value should always be a `Uint8Array`.
